qiskit/transpiler/passes/synthesis/plugin.py

```diff
@@ -0,0 +1,283 @@
+representing the synthesized circuit from that unitary matrix. Then to inform
+the Qiskit transpiler about what information is necessary for the pass there
+are several required property methods that need to be implemented,
+``supports_basis_gates``, ``supports_coupling_map``, and
+``supports_approximation_degree`` which return either ``True`` or ``False``
+depending on whether the plugin supports and/or requires that input to perform
+synthesis. An example plugin class would look something like::
+
+    from qiskit.transpiler.passes.synthesis import plugin
+
+    from qiskit_plugin_pkg.synthesis import generate_dag_circuit_from_matrix
+
+
+    class SpecialUnitarySynthesis(plugin.UnitarySynthesisPlugin):
+
+        @property
+        def supports_basis_gates(self):
+            return True
+
+        @property
+        def supports_coupling_map(self):
+            return False
+
+        @property
+        def supports_approximation_degree(self):
+            return False
+
+        @property
+        def supports_natural_direction(self):
+            return False
+
+        @property
+        def supports_pulse_optimize(self):
+            return False
+
+        @property
+        def supports_gate_lengths(self):
+            return False
+
+        @property
+        def supports_gate_errors(self):
+            return False
+
+        def run(self, unitary, **options):
+            basis_gates = options['basis_gates']
+            dag_circuit = generate_dag_circuit_from_matrix(unitary, basis_gates)
+            return dag_circuit
+
+If for some reason the available inputs to the
+:meth:`~qiskit.transpiler.passes.synthesis.plugin.UnitarySynthesisPlugin.run`
+method are insufficient please open an issue and we can discuss expanding the
+plugin interface with new opt-in inputs that can be added in a backwards
+compatible manner for future releases. Do note though that this plugin interface
+is considered stable and guaranteed to not change in a breaking manner. If
+changes are needed (for example to expand the available optional input options)
+it will be done in a way that will **not** require changes from existing
+plugins.
+
+The second step is to expose the
+:class:`~qiskit.transpiler.passes.synthesis.plugin.UnitarySynthesisPlugin` as
+a setuptools entry point in the package metadata. This is done by simply adding
+an ``entry_points`` entry to the ``setuptools.setup`` call in the ``setup.py``
+for the plugin package with the necessary entry points under the
+``qiskit.unitary_synthesis`` namespace. For example::
+
+    entry_points={
+        'qiskit.unitary_synthesis': [
+            'special = qiskit_plugin_pkg.module.plugin:SpecialUnitarySynthesis',
+        ]
+    },
+
+(note that the entry point ``name = path`` is a single string not a Python
+expression). There isn't a limit to the number of plugins a single package can
+include as long as each plugin has a unique name. So a single package can
+expose multiple plugins if necessary.
```

Comment:
Should we say something about the name `default` being built in?  Strange things might happen if anyone calls their plugin that.